### PR TITLE
make hapi support multiple gpus train and eval

### PR DESCRIPTION
**Pull request by LielinJiang**

### CODE REVIEW

File: distributed.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+            sys.stderr.write("server not ready, wait 3 sec to retry...\n")
+            sys.stderr.write("not ready endpoints:" + str(
+                not_ready_endpoints) + "\n")
+            sys.stderr.flush()
+            time.sleep(3)
+        else:
+            break
+
+
+def initCommunicator(program, rank, nranks, wait_port,
```
Comment (qingqing01) [inline distributed.py]: initCommunicator -> init_communicator

File: mnist.py [hunk 2]
```diff
@@ -97,88 +98,90 @@ def __init__(self):
...
+    # train_loader = fluid.io.xmap_readers(
+    #     lambda b: [np.array([x[0] for x in b]).reshape(-1, 1, 28, 28),
+    #                np.array([x[1] for x in b]).reshape(-1, 1)],
+    #     paddle.batch(fluid.io.shuffle(paddle.dataset.mnist.train(), 6e4),
+    #                  batch_size=FLAGS.batch_size, drop_last=True), 1, 1)
+    # val_loader = fluid.io.xmap_readers(
+    #     lambda b: [np.array([x[0] for x in b]).reshape(-1, 1, 28, 28),
+    #                np.array([x[1] for x in b]).reshape(-1, 1)],
+    #     paddle.batch(paddle.dataset.mnist.test(),
+    #                  batch_size=FLAGS.batch_size, drop_last=True), 1, 1)
```
Comment (heavengate) [inline mnist.py]: delete?

File: distributed.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        return
+
+    if isinstance(place, core.CUDAPlace):
+        communicator_prog = framework.Program()
+        initCommunicator(communicator_prog, strategy.local_rank, strategy.nranks, True,
+                         strategy.current_endpoint, strategy.trainer_endpoints)
+        exe = fluid.Executor(place)
+        exe.run(communicator_prog)
+    else:
+        # TODO(Yancey1989): add Gloo Parallel Context to support CPU parallel computation
```
Comment (heavengate) [inline distributed.py]: better to rewrite this

File: distributed.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        grad_vars = []
+        for param in self._layers.parameters():
+            # NOTE(zcd): The grad_ivar maybe no generated.
+            if param.trainable and param._grad_ivar():
+                g_var = param._grad_ivar()
+                grad_vars.append(g_var)
+                assert g_var not in grad_var_set
+                grad_var_set.add(g_var)
+
+        # FIXME(zcd): the type of the var should be LoDTensor, i.e
```
Comment (heavengate) [inline distributed.py]: same as above

File: distributed.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+from paddle.fluid.dygraph import to_variable, no_grad, layers
+from paddle.fluid.framework import Variable
+from paddle.fluid.executor import global_scope
+
+from paddle.fluid.dygraph.parallel import Env, DataParallel, ParallelStrategy
+from paddle.fluid.layers.collective import _c_allreduce, _c_allgather, _c_broadcast, _c_sync_comm_stream, _c_sync_calc_stream
+from paddle.fluid.io import BatchSampler, DataLoader
+
+
+class DistributedBatchSampler(BatchSampler):
```
Comment (qingqing01) [inline distributed.py]: 这里继承BatchSampler，但下面代码也没用到BatchSampler里的东西

File: progressbar.py [hunk 6]
```diff
@@ -2,6 +2,7 @@
 import time
 import numpy as np
 
+from distributed import get_local_rank
```
Comment (qingqing01) [inline progressbar.py]: 不用在progressbar中使用get_local_rank吧，在调用get_local_rank的地方使用即可，在callback中调用创建Progressbar和调用updates的地方就行吧

File: mnist.py [hunk 7]
```diff
@@ -97,88 +98,90 @@ def __init__(self):
...
         x = fluid.layers.flatten(x, axis=1)
         x = self._fc(x)
         return x
 
 
-def accuracy(pred, label, topk=(1, )):
-    maxk = max(topk)
-    pred = np.argsort(pred)[:, ::-1][:, :maxk]
-    correct = (pred == np.repeat(label, maxk, 1))
+class CustromMnistDataset(MnistDataset):
```
Comment (qingqing01) [inline mnist.py]: 下面不能直接用MnistDataset？

File: mnist.py [hunk 8]
```diff
@@ -97,88 +98,90 @@ def __init__(self):
...
 
 def main():
     @contextlib.contextmanager
     def null_guard():
         yield
 
-    guard = fluid.dygraph.guard() if FLAGS.dynamic else null_guard()
+    place = fluid.CUDAPlace(fluid.dygraph.parallel.Env().dev_id) \
+        if fluid.dygraph.parallel.Env().nranks > 1 else fluid.CUDAPlace(0)
+    guard = fluid.dygraph.guard(place) if FLAGS.dynamic else null_guard()
```
Comment (qingqing01) [inline mnist.py]: 动态图切换的guard里还需要传入palce?

File: mnist.py [hunk 9]
```diff
@@ -97,88 +98,90 @@ def __init__(self):
...
+    if fluid.dygraph.parallel.Env().nranks > 1:
+        prepare_context(place)
```
Comment (qingqing01) [inline mnist.py]: 多卡的使用对用户还是不够简化的样子

File: model.py [hunk 10]
```diff
@@ -350,6 +360,17 @@ def _run(self, inputs, labels=None):
         metric_states = restore_flatten_list(rets[num_loss:], metric_splits)
         metrics = []
         for metric, state in zip(self.model._metrics, metric_states):
+            # cut off padding size
+            if self.mode != 'train' and self.model._test_dataloader is not None and self._nranks > 1:
```
Comment (qingqing01) [inline model.py]: 这行超出了79个字符吧

File: model.py [hunk 11]
```diff
@@ -350,6 +360,17 @@ def _run(self, inputs, labels=None):
...
+                total_size = len(self.model._test_dataloader.dataset)
+                samples = state[0].shape[0]
```
Comment (qingqing01) [inline model.py]: 会不会存在state[0].shape[0]不是batch-size的情况，或者用inputs里的数据。可以加个TODO记录下。

File: model.py [hunk 12]
```diff
@@ -389,28 +411,45 @@ def _make_program(self, mode):
...
+                        
                     self.model._optimizer.minimize(self._loss_endpoint)
+            if self._nranks > 1 and mode != 'train' and self.model._test_dataloader is not None:
+                outputs = [distributed._all_gather(o, self._nranks) for o in outputs]
+                if mode != 'test':
+                    labels = [distributed._all_gather(l, self._nranks) for l in labels]
+                    
+            if mode != 'test':
+                for metric in self.model._metrics:
+                    metrics.append(to_list(metric.add_metric_op(outputs, labels)))
```
Comment (qingqing01) [inline model.py]: metric需要添加到backward的前面，虽然从逻辑上并没有错，但一般的习惯还是先forward->loss/metric->backward->opt吧

File: model.py [hunk 13]
```diff
@@ -735,6 +819,12 @@ def _run_one_epoch(data_loader, callbacks, mode):
                 'metrics_name': metrics_name,
             }
             for step, data in enumerate(data_loader):
+                if not fluid.in_dygraph_mode():
+                    data = data[0]
```
Comment (qingqing01) [inline model.py]: 静态图的时候 data是data[0] ?

File: distributed.py (hunk 5 above)
Comment (LielinJiang) [reply distributed.py]: 这个是因为凯鹏在dataloader里，assert了sampler必须为BatchSampler的子类

File: mnist.py [hunk 14]
```diff
@@ -97,88 +98,90 @@ def __init__(self):
...
+            train_loader = DataLoader(train_dataset, batch_sampler=train_sampler, places=place, 
+                                    feed_list=feed_list, num_workers=4, return_list=True)
+            val_sampler = DistributedBatchSampler(val_dataset, batch_size=FLAGS.batch_size)
+            val_loader = DataLoader(val_dataset, batch_sampler=val_sampler, places=place, 
+                                    feed_list=feed_list, num_workers=4, return_list=True)
+        else:
+            train_loader = DataLoader(train_dataset, batch_size=FLAGS.batch_size, places=place, 
+                                    feed_list=feed_list, num_workers=4, return_list=True)
+            val_loader = DataLoader(val_dataset, batch_size=FLAGS.batch_size, places=place, 
+                                    feed_list=feed_list, num_workers=4, return_list=True)
```
Comment (qingqing01) [inline mnist.py]: train_sampler = DistributedBatchSampler() if get_nranks() > 1 else None
val_sampler= ...

train_loader=DataLoader()
val_loader=DataLoader()

这样train_loader、val_loader个少写一个

File: mnist.py (hunk 8 above)
Comment (LielinJiang) [reply mnist.py]: 是的，在多卡的时候需要

File: mnist.py (hunk 14 above)
Comment (LielinJiang) [reply mnist.py]: 试了一下，由于定义DistributedBatchSampler，可行性不强，没法用一行的if else写

File: image_classification/utils.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    maxk = max(topk)
+    pred = np.argsort(pred)[:, ::-1][:, :maxk]
+    correct = (pred == np.repeat(label, maxk, 1))
+
+    batch_size = label.shape[0]
+    res = []
+    for k in topk:
+        correct_k = correct[:, :k].sum()
+        res.append(100.0 * correct_k / batch_size)
+    return res
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply image_classification/utils.py]: done, thanks

File: model.py [hunk 16]
```diff
@@ -348,6 +356,12 @@ def _run(self, inputs, labels=None):
         metric_states = restore_flatten_list(rets[num_loss:], metric_splits)
         metrics = []
         for metric, state in zip(self.model._metrics, metric_states):
+            # cut off padding size
+            if self.model._dataset is not None and self._nranks > 1:
+                total_size = len(self.model._dataset)
+                samples = state[0].shape[0]
+                if metric.count[0] + samples > total_size:
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply model.py]: done, thanks

File: model.py [hunk 17]
```diff
@@ -465,36 +512,52 @@ def train(self, inputs, labels=None):
...
         for metric in self.model._metrics:
-            metric_outs = metric.add_metric_op(outputs, to_list(labels))
+            metric_outs = metric.add_metric_op(to_list(outputs), to_list(labels))
             m = metric.update(*[to_numpy(m) for m in to_list(metric_outs)])
             metrics.append(m)
         return ([to_numpy(l) for l in losses], metrics) \
                 if len(metrics) > 0 else [to_numpy(l) for l in losses]
 
-    def eval(self, inputs, labels=None):
+    def eval(self, inputs, labels, device='CPU', device_ids=None):
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply model.py]: done, thanks

File: model.py [hunk 18]
```diff
@@ -465,36 +512,52 @@ def train(self, inputs, labels=None):
...
+            outputs = [distributed._all_gather(o, self._nranks) for o in to_list(outputs)]
+            labels = [distributed._all_gather(l, self._nranks) for l in labels]
         metrics = []
         for metric in self.model._metrics:
-            metric_outs = metric.add_metric_op(outputs, labels)
+            # cut off padding value.
+            if self.model._dataset is not None and self._nranks > 1:
+                total_size = len(self.model._dataset)
+                samples = outputs[0].shape[0]
+                if metric.count[0] + samples > total_size:
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply model.py]: done, thanks

File: model.py [hunk 19]
```diff
@@ -616,6 +684,7 @@ def prepare(self,
                 metrics=None,
                 inputs=None,
                 labels=None,
+                dataset=None,
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply model.py]: deleted

File: models/download.py [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+    with open(fullname, 'rb') as f:
+        for chunk in iter(lambda: f.read(4096), b""):
+            md5.update(chunk)
+    calc_md5sum = md5.hexdigest()
+
+    if calc_md5sum != md5sum:
+        logger.info("File {} md5 check failed, {}(calc) != "
+                    "{}(base)".format(fullname, calc_md5sum, md5sum))
+        return False
+    return True
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply models/download.py]: deleted

File: distributed.py (hunk 1 above)
Comment (LielinJiang) [reply distributed.py]: done, thanks

File: mnist.py (hunk 2 above)
Comment (LielinJiang) [reply mnist.py]: done, thanks

File: distributed.py (hunk 3 above)
Comment (LielinJiang) [reply distributed.py]: done, thanks

File: distributed.py (hunk 4 above)
Comment (LielinJiang) [reply distributed.py]: done, thanks

File: distributed.py [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+    .. note::
+        Dataset is assumed to be of constant size.
+        
+    Args:
+        data_source: this could be a `fluid.io.Dataset` implement
+                     or other python object which implemented
+                     `__len__` for BatchSampler to get sample
+                     number of data source.
+        batch_size(int): sample indice number in a mini-batch indices.
+        shuffle(bool): whther to shuffle indices order before genrate
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply distributed.py]: done, thanks

File: progressbar.py (hunk 6 above)
Comment (LielinJiang) [reply progressbar.py]: done, thanks

File: mnist.py (hunk 9 above)
Comment (LielinJiang) [reply mnist.py]: 试了很多方法，都有各种各样的问题。。后续再尝试一下

File: model.py (hunk 10 above)
Comment (LielinJiang) [reply model.py]: done, thanks

File: model.py (hunk 12 above)
Comment (LielinJiang) [reply model.py]: metric需要放在gather后面，gather需要放在fleet初始化后面。。。

File: model.py (hunk 13 above)
Comment (LielinJiang) [reply model.py]: 是的

File: mnist.py [hunk 22]
```diff
@@ -28,7 +28,8 @@
 
 from model import Model, CrossEntropy, Input
 from metrics import Accuracy
-
+from distributed import prepare_context, all_gather, Env, get_nranks, get_local_rank, DistributedBatchSampler, to_numpy
```
Comment (xyzhou-puck) [inline mnist.py]: 这里好复杂啊，需要import的东西太多了，而且抽象也不够。

File: mnist.py [hunk 23]
```diff
@@ -97,88 +98,79 @@ def __init__(self):
...
+
 
-    batch_size = label.shape[0]
-    res = []
-    for k in topk:
-        correct_k = correct[:, :k].sum()
-        res.append(100.0 * correct_k / batch_size)
-    return res
+    def __getitem__(self, idx):
+        return self.images[idx], [self.labels[idx]]
```
Comment (xyzhou-puck) [inline mnist.py]: dataset 能做到分布式和单机都融合地比较好吗？现在这块看起来比较复杂，而且需要用户手动搞的东西比较多。

File: model.py [hunk 24]
```diff
@@ -586,6 +666,10 @@ def __init__(self):
         self._optimizer = None
         self._device = None
         self._device_ids = None
+        self._optimizer = None
+        self._distributed_sampler = None
```
Comment (guoshengCS) [inline model.py]: `self._distributed_sampler` seems unused elsewhere, should we remove it

File: model.py [hunk 25]
```diff
@@ -458,17 +517,24 @@ def train(self, inputs, labels=None):
...
+            losses = self.model._loss_function(outputs, labels)
+            final_loss = fluid.layers.sum(losses)
+            final_loss.backward()
         self.model._optimizer.minimize(final_loss)
         self.model.clear_gradients()
         metrics = []
         for metric in self.model._metrics:
-            metric_outs = metric.add_metric_op(outputs, to_list(labels))
-            m = metric.update(* [to_numpy(m) for m in to_list(metric_outs)])
+            metric_outs = metric.add_metric_op(to_list(outputs), to_list(labels))
```
Comment (guoshengCS) [inline model.py]: Should we `_all_gather` outputs and labels before compute metirc, it seems these are all gathered in static graph train/eval/test modes,  while only eval/test modes in dygraph mode are gathered.

File: model.py [hunk 26]
```diff
@@ -800,12 +893,13 @@ def _run_one_epoch(data_loader, callbacks, mode):
...
                     res = metric.accumulate()
                     metrics.extend(to_list(res))
+                    
                 assert len(metrics_name) == len(metrics)
                 for k, v in zip(metrics_name, metrics):
                     logs[k] = v
 
                 logs['step'] = step
-                logs['batch_size'] = data[0].shape[0]
+                logs['batch_size'] = batch_size
```
Comment (guoshengCS) [inline model.py]: It seems  metrics result are gathered result, should `batch_size` also be summed up for multi-devices

File: model.py [hunk 27]
```diff
@@ -336,6 +346,19 @@ def _run(self, inputs, labels=None):
         metric_states = restore_flatten_list(rets[num_loss:], metric_splits)
         metrics = []
         for metric, state in zip(self.model._metrics, metric_states):
+            # cut off padding size
+            if self.mode != 'train' and self.model._test_dataloader is not None \
+                                    and self._nranks > 1:
+                total_size = len(self.model._test_dataloader.dataset)
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply model.py]: updated, thanks

File: distributed.py [hunk 28]
```diff
@@ -0,0 +1,297 @@
...
+
+def to_numpy(var):
+    assert isinstance(var, (Variable, fluid.core.VarBase)), "not a variable"
+    if isinstance(var, fluid.core.VarBase):
+        return var.numpy()
+    t = global_scope().find_var(var.name).get_tensor()
+    return np.array(t)
+
+
+def all_gather(input):
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply distributed.py]: removed, thanks

File: model.py (hunk 24 above)
Comment (LielinJiang) [reply model.py]: removed, thanks

File: model.py (hunk 26 above)
Comment (LielinJiang) [reply model.py]: done, thanks.
